roxie/ccd/ccdserver.cpp

```diff
@@ -1400,14 +1453,57 @@ class CRoxieServerActivity : implements CInterfaceOf<IRoxieServerActivity>, impl
         }
 #endif
         executeDependencies(parentExtractSize, parentExtract, 0);
-        if (input)
-            input->start(parentExtractSize, parentExtract, paused);
         ensureCreated();
         basehelper.onStart(parentExtract, NULL);
+        startInputs(parentExtractSize, parentExtract, paused);
         if (collectFactoryStatistics)
             factory->noteStarted();
+    }
+
+    virtual void startInputs(unsigned parentExtractSize, const byte *parentExtract, bool paused)
+    {
+        if (input)
+            input->start(parentExtractSize, parentExtract, paused);
         startJunction(junction);
     }
+    
+    virtual RoxieSourceCharacteristics getSourceCharacteristics() const override 
+    {
+        // Note - there's a possibility that two threads arriving here very close together will result in parent getSourceCharacteristics being called twice,
+        // but that is harmless. Not even sure this is ever called from multiple threads
+        if (!sourceCharacteristicsCalculated)
+        {
+            RoxieSourceCharacteristics characteristics = getActivityCharacteristics() | getAllInputsCharacteristics();
+            if (dependencies.ordinality())
+            {
+                characteristics |= RSC::hasDependencies;
+                ForEachItemIn(idx, dependencies)
+                {
+                    RoxieSourceCharacteristics dependCharacteristics = dependencies.item(idx).getSourceCharacteristics();
+                    if (hasMask(dependCharacteristics, RSC::urgentStart|RSC::hasRowLatency|RSC::slowDependencies))
+                        characteristics |= RSC::slowDependencies;
+                }
+            }
+            cachedSourceCharacteristics = characteristics;
+            sourceCharacteristicsCalculated = true;
+            if (traceActivityCharacteristics && characteristics!=RSC::none)
+                CTXLOG("Activity %s characteristics: %s", factory->describe().c_str(), getText(characteristics).c_str());
+            stats.addStatistic(StEnumActivityCharacteristics, (unsigned) characteristics);
```

Comment:
Rather than storing it as a statistic, I would add code in CRoxieServerActivity::gatherStats()

     if (sourceCharacteristicsCalculated)
         merged.mergeStatistic(StEnumActivityCharacteristics, (unsigned) cachedSourceCharacteristics);

to avoid saving the values twice.